File: xllm/core/layers/common/activation.cpp
/* Copyright 2025 The xLLM Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    https://github.com/jd-opensource/xllm/blob/main/LICENSE

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "activation.h"

#include "kernels/ops_api.h"

namespace xllm {
namespace layer {

ActivationImpl::ActivationImpl(const std::string& act_mode, bool is_gated)
    : act_mode_(act_mode), is_gated_(is_gated) {}

void ActivationImpl::forward(torch::Tensor& input, torch::Tensor& output) {
  xllm::kernel::ActivationParams activation_params;
  activation_params.input = input;
  activation_params.output = output;
  activation_params.act_mode = act_mode_;
  activation_params.is_gated = is_gated_;
  xllm::kernel::active(activation_params);
}

}  // namespace layer
}  // namespace xllm

File: xllm/python/kernels_cuda/activation.py
# Copyright 2026 The xLLM Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://github.com/xLLM-AI/xllm/blob/main/LICENSE
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""CUDA activation kernels."""

from __future__ import annotations

import torch

silu_and_mul = torch.ops.xllm_ops.silu_and_mul


@torch.library.custom_op("xllm_triton::silu_and_mul", mutates_args=())
def _silu_and_mul_triton(value: torch.Tensor) -> torch.Tensor:
    """Run the Triton gated SiLU kernel as one graph node.

    Kept alongside the C++ operator so a model graph can pick the Triton path
    explicitly through ``torch.ops.xllm_triton.silu_and_mul``.
    """
    from .triton.silu_and_mul import (
        silu_and_mul as triton_silu_and_mul,
    )

    return triton_silu_and_mul(value)


@_silu_and_mul_triton.register_fake
def _silu_and_mul_triton_fake(value: torch.Tensor) -> torch.Tensor:
    shape = list(value.shape)
    shape[-1] //= 2
    return value.new_empty(shape)


__all__ = ["silu_and_mul"]
